### PR TITLE
Forward: introduce `Opt` builder for creating plugin instance.

### LINKED ISSUE #2182: Expose support for forward's properties settings
Hi, one of my plugin may require reuse the forward plugin's functionality. Firstly I try to open a nested block for forward:

```
. {
  myplugin {
    myplugin_settings xxx
    forward . 1.1.1.1 {
      duration 10
    }
  }
}
```

But with no luck, caddyfile's dispenser seems have no support for this nested usage (and I do agree it's a bad style for config).

Another approach looks like:

```
. {
  myplugin {
    myplugin_settings_foo foo
    forward_expire DURATION
    // ... other properties for forward, prefixed with `forward`
  }
}
```

```go
type MyPlugin struct {
  Forward forward.Forward
}

func setup(c *caddy.Controller) *MyPlugin {
  p := New()

  // do parse work here
  parseStanza(c, p)

  return c
}

// so in parseStanza we can setup the forward plugin with exposed setters:
func parseStanza(c *caddy.Controller, p *MyPlugin) {
  // reading from caddyfile.Dispenser
  switch dispenser.Val() {
    case "forward_duration":
      p.Forward.SetDuration(parsedDuration)
    case "other_property":
      // ...
  }
}
```

Some may ask why not just fall through and let the chain handle this case, but to my usage, there is a need for altering the chain flowing direction, like a `if` statement, generally speaking, the flow would look like that:

```
. {
  if {
    condition_to_check
    perform_when_condition_meet
  }

  fallthrough_plugin
}
```

Any suggestions for this change? If you agree with this modification, I can submit a pr.

### PR DESCRIPTION
**Pull request by bcho**
fixes: #2182

### 1. Why is this pull request needed and what does it do?

This pull request adds a simple implementation for #2182, which decouples the forward plugin building logic, and enables reusing later.

### 2. Which issues (if any) are related?

#2182 

### 3. Which documentation changes (if any) need to be made?

forward plugin's `New` method may be marked as deprecated

### CODE REVIEW

File: plugin/forward/forward.go [hunk 1]
```diff
@@ -21,6 +23,72 @@ import (
...
+		TLSConfig:           new(tls.Config),
+		Maxfails:            2,
+		Expire:              defaultExpire,
+		HealthCheckInterval: hcInterval,
+		SelectPolicy:        new(random),
+	}
+}
+
+// Build creates a forward instance.
+func (o Opt) Build() *Forward {
```
Comment (johnbelamaric) [inline plugin/forward/forward.go]: This should be (o *Opt) or it will make a copy.

File: plugin/forward/setup.go [hunk 2]
```diff
@@ -89,51 +88,34 @@ func parseForward(c *caddy.Controller) (*Forward, error) {
...
-			f.proxies[i].SetTLSConfig(f.tlsConfig)
-		}
-		f.proxies[i].SetExpire(f.expire)
-	}
-	return f, nil
+	return o.Build(), nil
 }
 
-func parseBlock(c *caddyfile.Dispenser, f *Forward) error {
+func parseBlock(c *caddyfile.Dispenser, opt *Opt) error {
```
Comment (johnbelamaric) [inline plugin/forward/setup.go]: Why name it "opt" here and "o" everywhere else?

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
A few minor tweaks but otherwise lgtm. I know thought that @miekg has had concerns in the past about exposing this sort of thing in a plugin, rather than moving to a common pkg.

**Comment by bcho** [comment]
@johnbelamaric changes updated, PTAL.

> concerns in the past about exposing this sort of thing in a plugin, rather than moving to a common pkg

Yep, I have the same concern too, but I don't find any implementations applying this way. @miekg any suggestions here?

**Comment by miekg** [comment]
How is this different than just exposing the bits you need?

The problem with the approach is this now puts a requirement on this pluign to not break this api - there was never a thing like this before, so it's a pretty big change.

What's the usecase here and why can this not be solved by the various Lookup functions that are sort of the entry point for other plugin into this one?
(even with something like this in place it will break at some point, because there isn't any inherit guarantee being agreed here)

**Comment by bcho** [comment]
Hi @miekg 

> What's the usecase here

The main use case is to allow other plugins to embed forward plugin's capacity (basically, to use `Forward.ServeDNS`).

For now, `Forward` holds some internal fields and I think it can be instantiated by an `Opt` builder. This builder actually reduces some dependencies between the caller and the callee IMO. Of course, this kind of usage requires the API to be stable, as the use case is just want to create dependencies between plugins.

But I come up with a better way, like just copy the `ServeDNS` logic into my plugin like that:

```go
import (
  "github.com/coredns/coredns/plugin/forward"
)

func (p MyPlugin) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) {
  if fallbackToForward(ctx, w, r) {
    p.serveDNSWithForward(ctx, w, r)
  }
}

func (p MyPlugin) serveDNSWithForward(// ...) {
  var proxies []*forward.Proxy
  // using proxy as the same way forward does
}
```

So I think it's ok to close this pull request by now.

**Pull request closed without merging**